presentation-compiler/src/main/scala/org/alephium/ralph/lsp/pc/search/gototype/GoToTypeCodeProvider.scala

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Alephium
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package org.alephium.ralph.lsp.pc.search.gototype
+
+import org.alephium.ralph.lsp.access.compiler.ast.Tree
+import org.alephium.ralph.lsp.access.compiler.message.SourceIndexExtra.SourceIndexExtension
+import org.alephium.ralph.lsp.pc.search.CodeProvider
+import org.alephium.ralph.lsp.pc.sourcecode.{SourceCodeState, SourceLocation}
+import org.alephium.ralph.lsp.pc.workspace.WorkspaceState
+import org.alephium.ralph.lsp.utils.log.{ClientLogger, StrictImplicitLogging}
+import org.alephium.ralph.lsp.utils.Node
+import org.alephium.ralph.Ast
+
+case object GoToTypeCodeProvider extends CodeProvider[SourceCodeState.Parsed, Unit, SourceLocation.GoToType] with StrictImplicitLogging {
+
+  /**
+   * Performs a search operation at the cursor index within the source-code of a workspace.
+   *
+   * @param cursorIndex    The index (character offset) in the source code representing the cursor position.
+   * @param sourceCode     The source code state where the search is executed.
+   * @param workspace      The workspace state where the source-code is located.
+   * @param searchSettings Provider-specific settings.
+   * @return Search results.
+   */
+  override def search(
+      cursorIndex: Int,
+      sourceCode: SourceCodeState.Parsed,
+      workspace: WorkspaceState.IsSourceAware,
+      searchSettings: Unit
+    )(implicit logger: ClientLogger): Iterator[SourceLocation.GoToType] =
+    sourceCode.astStrict.statements.find(_.index contains cursorIndex) match {
+      case Some(statement) =>
+        statement match {
+          case _: Tree.Import =>
+            Iterator.empty
+
+          case tree: Tree.Source =>
+            tree.rootNode.findLast(_.sourceIndex.exists(_ contains cursorIndex)) match {
+              case Some(node @ Node(ident: Ast.Ident, _)) =>
+                GoToTypeIdent(
+                  node = node.upcast(ident),
+                  workspace = workspace
+                ).iterator
+
+              case Some(Node(data, _)) =>
+                logger.error(s"GoToType not implemented for ${data.getClass.getName}. Index: $cursorIndex. FileURI: ${sourceCode.fileURI}")
```

Comment:
That would be nice at some point if we could return those errors to users.

So instead of seeing: `No locations found` (at least on my nvim plugin)

We could have an error message, but I'm not 100% sure it's a good idea as the protocol says only:
`error: code and message set in case an exception happens during the definition request`
![image](https://github.com/user-attachments/assets/dd93922e-cf54-474c-b89e-eb0f06ad08a7)

Nothing urgent, it just popped in my mind

Reply:
Yep, good idea 👍🏼. In the beginning, there weren't any cases where `CodeProvider`s had to report/return an error. But later on they do, and they also need to return a `Future` for `SoftAST`. I think they could all be address https://github.com/alephium/ralph-lsp/issues/453.